cli/internal/daemon/connector/connector.go

```diff
@@ -0,0 +1,320 @@
+	_socketPollTimeout  = 1 * time.Second
+)
+
+// killLiveServer tells a running server to shut down. This method is also responsible
+// for closing this connection
+func (c *Connector) killLiveServer(ctx context.Context, client *Client, serverPid int) error {
+	defer func() { _ = client.Close() }()
+
+	_, err := client.Shutdown(ctx, &turbodprotocol.ShutdownRequest{})
+	if err != nil {
+		c.Logger.Error(fmt.Sprintf("failed to shutdown running daemon. attempting to force it closed: %v", err))
+		return c.killDeadServer(serverPid)
+	}
+	// Wait for the server to gracefully exit
+	deadline := time.After(_shutdownDeadline)
+	for {
+		lockFile, err := lockfile.New(c.PidPath.ToString())
+		if err != nil {
+			return err
+		}
+		owner, err := lockFile.GetOwner()
+		if os.IsNotExist(err) {
+			return nil
+		} else if err != nil {
+			return err
+		} else if owner.Pid == serverPid {
+			// We're still waiting for the server to shut down
+			select {
+			case <-deadline:
+				c.Logger.Error(fmt.Sprintf("daemon did not exit after %v, attempting to force it closed", _shutdownDeadline.String()))
+				return c.killDeadServer(serverPid)
+			case <-time.After(_shutdownPollInterval):
+				// loop around and check again
+			}
+		}
+	}
+}
+
+func (c *Connector) killDeadServer(pid int) error {
+	// currently the only error that this constructor returns is
+	// in the case that you don't provide an absolute path.
+	// Given that we require an absolute path as input, this should
+	// hopefully never happen.
+	lockFile, err := lockfile.New(c.PidPath.ToString())
+	if err != nil {
+		return err
+	}
+	process, err := lockFile.GetOwner()
+	if err == nil {
+		// Check that this is the same process that we failed to connect to.
+		// Otherwise, loop around again and start with whatever new process has
+		// the pid file.
+		if process.Pid == pid {
+			// we have a process that we need to kill
+			// TODO(gsoltis): graceful kill? the process is already not responding to requests,
+			// but it could be in the middle of a graceful shutdown. Probably should let it clean
+			// itself up, and report an error and defer to a force-kill by the user
+			if err := process.Kill(); err != nil {
+				return err
+			}
+		}
+		return nil
+	} else if errors.Is(err, os.ErrNotExist) {
+		// There's no pid file. Someone else killed it. Loop around and try again
+		return nil
+	} else if errors.Is(err, lockfile.ErrDeadOwner) {
+		// The daemon crashed. report an error to the user
+		return err
+	}
+	return err
+}
+
+// Connect attempts to create a connection to a turbo daemon.
+// Retries and daemon restarts are built in. If this fails,
+// it is unlikely to succeed after an automated retry.
+func (c *Connector) Connect(ctx context.Context) (*Client, error) {
+	client, err := c.connectInternal(ctx)
+	if err != nil {
+		return nil, c.wrapConnectionError(err)
+	}
+	return client, nil
+}
+
+func (c *Connector) connectInternal(ctx context.Context) (*Client, error) {
```

Comment:
This function is doing a lot. First PID, then Sock, then Hello. I don't know how well it works for scoping reasons, but if it's not too bad can we break those into separate named tasks?

Or verbose comments on the focus switches. "Okay, we've got the PID lock, so next we need to ..."

Reply:
Pulled out the `getOrStartDaemon` part, as well as added some more detailed comments.